hashicorp/waypoint
# Automatically init projects with data sources

## Why the change?

For projects that are configured for GitOps but *not* for automatic polling, we currently require users to step out of the Web UI to the CLI to initialize the project before they can continue. This PR allows Waypoint to remotely initialize this type of gitops project, and to display errors encountered during remote initialization.

## What’s the plan?

- [x] Add `InitOp` and `InitResult` protos
- [x] Figure out when/how to schedule an InitOp
- [x] Figure out what to display in the UI when an InitOp is (potentially) happening
- [x] Figure out what to display in the UI if an InitOp fails (if anything, maybe fall back to CLI instructions)
- [x] Spruce up error code block (ideally use something that can be swapped out with Greg’s upcoming log output component)

## What does it look like?

### Happy Path (Light Mode)

https://user-images.githubusercontent.com/34030/131908007-3b7fddd5-fa81-4fe5-b751-119ca16056b3.mp4

### Happy Path (Dark Mode)

https://user-images.githubusercontent.com/34030/131915715-5d5bc76b-6e31-4e4b-9409-c5ec3e7be5cd.mp4

### Sad Path (Light Mode)

https://user-images.githubusercontent.com/34030/131916341-25d04de8-a57a-40a6-a215-44a3d1aa0011.mp4

### Sad Path (Dark Mode)

https://user-images.githubusercontent.com/34030/131916373-b9f798c6-03ec-4201-8a17-050f2ef21c5d.mp4

## How do I test it?

1. Check out the branch
2. Compile waypoint (`make static-assets` has been run ahead of time to ease testing)
3. Install to your platform of choice (with runners)
4. Visit the browser UI (`waypoint ui -authenticate`)
5. Create a new project with a valid Git source, but with “automatic deployments on changes” disabled
6. Verify that you see the “Initializing from…” messages
7. Verify that the project’s apps automatically show up in the UI
8. Try creating a project with a broken data source or broken config
9. Verify that you see an actionable error message in the UI
